metric/producer/manager.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package producer
+
+import (
+	"sync"
+)
+
+type manager struct {
+	mu       sync.RWMutex
+	producers []Producer
+}
+
+var prodMgr *manager
+var once sync.Once
+
+func getManager() *manager {
+	once.Do(func() {
+		prodMgr = &manager{}
+	})
+	return prodMgr
+}
+// Add adds the producer to the manager if it is not already present.
+func Add(producer Producer) {
+	pm := getManager()
+	pm.add(producer)
+}
+
+// Delete deletes the producer from the manager if it is present.
+func Delete(producer Producer) {
+	pm := getManager()
+	pm.delete(producer)
+}
+
+// GetAll returns all producer registered with the manager. It is typically
+// used by exporter to read metrics from producers.
+func GetAll() []Producer{
+	pm := getManager()
+	return pm.producers
```

Comment:
Consider returning a copy of producer slice, otherwise this may introduce synchronization issue. In Java we always return [an unmodifiable snapshot](https://github.com/census-instrumentation/opencensus-java/blob/master/impl_core/src/main/java/io/opencensus/implcore/metrics/export/MetricProducerManagerImpl.java#L32).

Reply:
slices are unmodifiable implicitly. I have added a test for it.

Reply:
That is not true:
I think this is a good article to read https://medium.com/@nitishmalhotra/uh-ohs-in-go-slice-of-pointers-c0a30669feee

Reply:
I gave you the wrong link. The idea is that Slice is a struct {len, cap, pointerToData}. If passed by value all of these will get copied but the underlying memory where  pointerToData points is shared between multiple copies of the slice.

Reply:
I got it. It is NOT immutable. I verified by a different test.

Reply:
I will change the implementation.